### PR TITLE
Create Vendor & Location entities

### PR DESCRIPTION
**Pull request by jioter**
Add vendor entity and OneToMany relation to Location entity

### CODE REVIEW

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    @Column(name = "id")
+    private Long id;
+
+    @Column(name = "country", length = 50, unique = true, nullable = false)
+    private String country;
+
+    @Column(name = "city", length = 50, nullable = false)
+    private String city;
+
+    @CreationTimestamp
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: it is better to use spring data annotation
org.springframework.data.annotation.CreatedDate

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    @EqualsAndHashCode.Exclude
+    @Column(name = "created_at", nullable = false)
+    private LocalDateTime createdAt;
+
+    @UpdateTimestamp
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: it is better to use spring data annotation
org.springframework.data.annotation.LastModifiedDate

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    @Column(name = "country", length = 50, unique = true, nullable = false)
+    private String country;
+
+    @Column(name = "city", length = 50, nullable = false)
+    private String city;
+
+    @CreationTimestamp
+    @EqualsAndHashCode.Exclude
+    @Column(name = "created_at", nullable = false)
+    private LocalDateTime createdAt;
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: the name "created" is enough

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Column(name = "country", length = 50, unique = true, nullable = false)
+    private String country;
+
+    @Column(name = "city", length = 50, nullable = false)
+    private String city;
+
+    @CreationTimestamp
+    @EqualsAndHashCode.Exclude
+    @Column(name = "created_at", nullable = false)
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: this column should be initialized only during first creation, and then it should never be updated.

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+@NoArgsConstructor
+@EqualsAndHashCode
+@Table(name = "location")
+public class Location {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id")
+    private Long id;
+
+    @Column(name = "country", length = 50, unique = true, nullable = false)
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: not a unique

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+@Entity
+@Data
+@NoArgsConstructor
+@EqualsAndHashCode
+@Table(name = "location")
+public class Location {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id")
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: rename to "loc_id"

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    private Long id;
+
+    @Column(name = "country", length = 50, unique = true, nullable = false)
+    private String country;
+
+    @Column(name = "city", length = 50, nullable = false)
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: rename to loc_city

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    private LocalDateTime modifiedAt;
+
+    @OneToMany(mappedBy = "location")
+    private List<Vendor> vendors;
+
+
+    public Location(String country, String city, LocalDateTime createdAt, LocalDateTime modifiedAt) {
+        this.country = country;
+        this.city = city;
+        this.createdAt = createdAt;
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: remove created & modified from constructor

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    @Column(name = "created_at", nullable = false)
+    private LocalDateTime createdAt;
+
+    @UpdateTimestamp
+    @EqualsAndHashCode.Exclude
+    @Column(name = "modified_at", nullable = false)
+    private LocalDateTime modifiedAt;
+
+    @OneToMany(mappedBy = "location")
+    private List<Vendor> vendors;
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/location/entity/Location.java:46]: Exclude from equals & hashcode

File: src/main/java/com/exadel/discountwebapp/vendor/entity/Vendor.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+@Data
+@Table(name = "vendor")
+@NoArgsConstructor
+@EqualsAndHashCode
+public class Vendor {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(name = "title", length = 55, nullable = false)
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/vendor/entity/Vendor.java]: vn_title

File: src/main/java/com/exadel/discountwebapp/vendor/entity/Vendor.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    private LocalDateTime created;
+
+    @EqualsAndHashCode.Exclude
+    @LastModifiedDate
+    @Column(name = "vn_modified")
+    private LocalDateTime modified;
+
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
+    @JoinColumn(name = "location_id")
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/vendor/entity/Vendor.java]: use the same name as in locaiton table "loc_id"

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java (hunk 1 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: @dmitriychernov7 changed to annotation @СreateDate

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java (hunk 2 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: @dmitriychernov7 edited to @LastModifiedDate

Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: @dmitriychernov7 edited to @LastModifiedDate

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java (hunk 3 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: @dmitriychernov7 amended

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java (hunk 4 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: @dmitriychernov7 added constraint upgradable=false

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java (hunk 5 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: @dmitriychernov7 unique constraint removed

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java (hunk 6 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: @dmitriychernov7 done

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java (hunk 7 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: @dmitriychernov7 done

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java (hunk 8 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/entity/Location.java]: @dmitriychernov7 added exclude from equals-hashcode annotation

File: src/main/java/com/exadel/discountwebapp/location/entity/Location.java (hunk 9 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/location/entity/Location.java:46]: added

File: src/main/java/com/exadel/discountwebapp/vendor/entity/Vendor.java (hunk 10 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/vendor/entity/Vendor.java]: done

File: src/main/java/com/exadel/discountwebapp/vendor/entity/Vendor.java (hunk 11 above)
Comment (jioter) [reply src/main/java/com/exadel/discountwebapp/vendor/entity/Vendor.java]: chenged

**Comment by jioter** [comment]
Can I do merge to master or do use have some more improvement comments?

**Review by ahlukhouski2 — APPROVED** [review]

**Review by Bozhokmaria — APPROVED** [review]

**Review by VolodymyrBushko — APPROVED** [review]

**Pull request merged**